File: clients/Go/model/camera.go
package model

import "io"
import . "aicup2020/stream"

type Camera struct {
    Center Vec2Float32
    Rotation float32
    Attack float32
    Distance float32
    Perspective bool
}
func NewCamera(center Vec2Float32, rotation float32, attack float32, distance float32, perspective bool) Camera {
    return Camera {
        Center: center,
        Rotation: rotation,
        Attack: attack,
        Distance: distance,
        Perspective: perspective,
    }
}
func ReadCamera(reader io.Reader) Camera {
    result := Camera {}
    result.Center = ReadVec2Float32(reader)
    result.Rotation = ReadFloat32(reader)
    result.Attack = ReadFloat32(reader)
    result.Distance = ReadFloat32(reader)
    result.Perspective = ReadBool(reader)
    return result
}
func (value Camera) Write(writer io.Writer) {
    value.Center.Write(writer)
    WriteFloat32(writer, value.Rotation)
    WriteFloat32(writer, value.Attack)
    WriteFloat32(writer, value.Distance)
    WriteBool(writer, value.Perspective)
}

File: clients/Ruby/model/camera.rb
require_relative 'vec2_float'
class Camera
    attr_accessor :center
    attr_accessor :rotation
    attr_accessor :attack
    attr_accessor :distance
    attr_accessor :perspective
    def initialize(center, rotation, attack, distance, perspective)
        @center = center
        @rotation = rotation
        @attack = attack
        @distance = distance
        @perspective = perspective
    end
    def self.read_from(stream)
        center = Vec2Float.read_from(stream)
        rotation = stream.read_float()
        attack = stream.read_float()
        distance = stream.read_float()
        perspective = stream.read_bool()
        Camera.new(center, rotation, attack, distance, perspective)
    end
    def write_to(stream)
        @center.write_to(stream)
        stream.write_float(@rotation)
        stream.write_float(@attack)
        stream.write_float(@distance)
        stream.write_bool(@perspective)
    end
end
